### PR TITLE
fix: use in memory provider for e2e suites

### LINKED ISSUE #712: Use in-memory provider for e2e suites
Currently, we use flagd for e2e testing in this repo (see following links). Now that we have in-memory providers for both web and server, we should remove the dependency on flagd and use the in-memory providers for these gherkin suites.

Remove this file: https://github.com/open-feature/js-sdk/blob/main/packages/server/e2e/step-definitions/setup.ts and this file: https://github.com/open-feature/js-sdk/blob/main/packages/client/e2e/step-definitions/setup.ts and configure the in-memory provider in the evaluation.spec gherkin implementations themselves ([here](https://github.com/open-feature/js-sdk/blob/a3799e1b658bafca5ea11528399e7560ef43d3e6/packages/server/e2e/step-definitions/evaluation.spec.ts#L22) and [here](https://github.com/open-feature/js-sdk/blob/a3799e1b658bafca5ea11528399e7560ef43d3e6/packages/client/e2e/step-definitions/evaluation.spec.ts#L20)).

### PR DESCRIPTION
**Pull request by znonthedev**
## This PR
Removes setup using Flagd Provider and uses In-Memory Provider

Fixes #712

### CODE REVIEW

**Comment by znonthedev** [comment]
I had issues running the e2e locally

```
Running one project: client-e2e
 FAIL   client-e2e  packages/client/e2e/step-definitions/evaluation.spec.ts
  ● Test suite failed to run

    packages/client/e2e/step-definitions/evaluation.spec.ts:10:63 - error TS2307: Cannot find module '../..' or its corresponding type declarations.

    10 import { OpenFeature, ProviderEvents, InMemoryProvider } from '../..';
```

I saw a PR related to this https://github.com/open-feature/js-sdk/pull/471

I changed the path locally to '../../src/' to get it working

**Review by lukas-reining — CHANGES_REQUESTED** [review]
Hey @znonthedev,
thanks for submitting that change, this looks good!

The thing left would be to remove the `flagd-testbed` from the GitHub Action.

https://github.com/open-feature/js-sdk/blob/main/.github/workflows/pr-checks.yaml#L68-L72

**Review by toddbaert — CHANGES_REQUESTED** [review]
Thanks @znonthedev ! This is a stop forward. The e2e tests are still failing though because you'll need to set up the `InMemoryProvider` to have the expected flags (you can do this in the constructor).

```
FAIL server-e2e packages/server/e2e/step-definitions/evaluation.spec.ts
  Flag evaluation
    ✕ Resolves boolean value (9 ms)
    ✕ Resolves string value (2 ms)
    ✕ Resolves integer value (1 ms)
    ✕ Resolves float value
    ✕ Resolves object value
    ✕ Resolves boolean details (1 ms)
    ✕ Resolves string details (1 ms)
    ✕ Resolves integer details
    ✕ Resolves float details
    ✕ Resolves object details (1 ms)
    ✕ Resolves based on context (1 ms)
    ✓ Flag not found (1 ms)
    ✕ Type error (2 ms)
```

The flagd test-bed has these flags configured. All the flags are defined in the gherkin suite [here](https://openfeature.dev/specification/appendix-b). The setup would like something like this (I've only included the first flag)

```ts
    OpenFeature.setProvider(
      new InMemoryProvider({
        'boolean-flag': {
          variants: {
            on: true,
            off: false,
          },
          disabled: false,
          defaultVariant: 'on',
        },
        // ...other flags...
      }),
    );
```

**Comment by znonthedev** [comment]
@lukas-reining  @toddbaert . Okay will make the necessary changes

**Comment by znonthedev** [comment]
@lukas-reining @toddbaert I have updated the PR with the requested changes. Please review. Why don't we remove the workflow and docker file related to flagd-testbed? I assume we don't need it anymore

**Comment by toddbaert** [comment]
> Why don't we remove the workflow and docker file related to flagd-testbed? I assume we don't need it anymore

Do you mean the `service` in the workflow (I assume so, since you removed it)? If so, yes! That's what @lukas-reining was suggest as well, I think.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -30,11 +30,7 @@ Run tests with `npm test`.
 
 ### End-to-End Tests
 
-The continuous integration runs a set of [gherkin e2e tests](https://github.com/open-feature/test-harness/blob/main/features/evaluation.feature) using [`flagd`](https://github.com/open-feature/flagd). These tests run with the "e2e" npm script. If you'd like to run them locally, you can start the flagd testbed with
-```
-docker run -p 8013:8013 ghcr.io/open-feature/flagd-testbed:latest
-```
-and then run
+The continuous integration runs a set of [gherkin e2e tests](https://github.com/open-feature/test-harness/blob/main/features/evaluation.feature) using in-memory provider. These tests run with the "e2e" npm script. If you'd like to run
```
Comment (toddbaert) [inline CONTRIBUTING.md:33]: Thanks for updating this, but I think you missed a few final words.

File: packages/client/e2e/step-definitions/flags-config.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  },
+  'timestamp-flag': {
+    disabled: false,
+    variants: {
+      past: -1,
+      future: 1,
+      none: 0,
+    },
+    defaultVariant: 'none',
+  },
```
Comment (toddbaert) [inline packages/client/e2e/step-definitions/flags-config.ts:71]: I could be wrong, but I don't think you need this one. This is a flagd-specific feature that doesn't need testing here.

File: packages/server/e2e/step-definitions/flags-config.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  },
+  'timestamp-flag': {
+    disabled: false,
+    variants: {
+      past: -1,
+      future: 1,
+      none: 0,
+    },
+    defaultVariant: 'none',
+  },
```
Comment (toddbaert) [inline packages/server/e2e/step-definitions/flags-config.ts:71]: Same as https://github.com/open-feature/js-sdk/pull/740/files#r1440652819

**Review by toddbaert — COMMENTED** [review]
@znonthedev Great work! everything passes, I just have a couple small nits, then I'm ready to approve.

Would you like an invite to the org? It comes with no obligation, but it's the first step if you'd like to be more involved.

File: packages/client/e2e/step-definitions/flags-config.ts (hunk 2 above)
Comment (znonthedev) [reply packages/client/e2e/step-definitions/flags-config.ts:71]: You're right. This flag config is not needed. I reused them from flagd testing-flags.json file. Removed it

File: packages/server/e2e/step-definitions/flags-config.ts (hunk 3 above)
Comment (znonthedev) [reply packages/server/e2e/step-definitions/flags-config.ts:71]: Removed the unwanted config

File: CONTRIBUTING.md (hunk 1 above)
Comment (znonthedev) [reply CONTRIBUTING.md]: Its a typo from my end. I forgot to stage the corrected version.

**Comment by znonthedev** [comment]
> > Why don't we remove the workflow and docker file related to flagd-testbed? I assume we don't need it anymore
> 
> Do you mean the `service` in the workflow (I assume so, since you removed it)? If so, yes! That's what @lukas-reining was suggesting as well, I think.

Not only the service but also the files within the test-harness submodule. I might be missing some context here. Are we using the flagd-testbed container for testing other SDKs as well?

**Comment by znonthedev** [comment]
> @znonthedev Great work! everything passes, I just have a couple small nits, then I'm ready to approve.
> 
> Would you like an invite to the org? It comes with no obligation, but it's the first step if you'd like to be more involved.



> @znonthedev Great work! everything passes, I just have a couple small nits, then I'm ready to approve.
> 
> Would you like an invite to the org? It comes with no obligation, but it's the first step if you'd like to be more involved.

@toddbaert Thanks a lot for the positive feedback!  I'm glad everything is passing smoothly. I appreciate the opportunity to contribute.
I'd be honoured to receive an invite to the org! While my availability may vary, I'm eager to contribute more and be part of the team.
Looking forward to the next steps, and thanks again for the opportunity!

**Comment by toddbaert** [comment]
> Not only the service but also the files within the test-harness submodule. I might be missing some context here. Are we using the flagd-testbed container for testing other SDKs as well?

I'm not 100% sure whether we need that submodule now, actually. I will take a quick look and follow up. Good catch...

**Comment by znonthedev** [comment]
@toddbaert Okay, then I think no changes are needed for the shared test harness files. However, since we removed the flagd-testbed dependency in this repo, we could remove the submodule itself from this repo. I may be wrong.

**Review by beeme1mr — APPROVED** [review]

**Comment by beeme1mr** [comment]
@znonthedev Todd won't be back until Monday. I'm going to leave this open until he has a chance to approve it. I think we should be good but I would like to confirm with him. Thanks for your effort with this.

File: packages/server/e2e/step-definitions/evaluation.spec.ts [hunk 4]
```diff
@@ -8,7 +8,8 @@ import {
   StandardResolutionReasons,
   ProviderEvents,
 } from '@openfeature/core';
-import { OpenFeature } from '../..';
+import { OpenFeature, InMemoryProvider } from '../..';
```
Comment (toddbaert) [inline packages/server/e2e/step-definitions/evaluation.spec.ts:11]:
```suggestion
import { OpenFeature, InMemoryProvider } from '../../src';
```

File: packages/client/e2e/step-definitions/evaluation.spec.ts [hunk 5]
```diff
@@ -7,7 +7,8 @@ import {
   ResolutionDetails,
   StandardResolutionReasons,
 } from '@openfeature/core';
-import { OpenFeature, ProviderEvents } from '../..';
+import { OpenFeature, ProviderEvents, InMemoryProvider } from '../..';
```
Comment (toddbaert) [inline packages/client/e2e/step-definitions/evaluation.spec.ts:10]:
```suggestion
import { OpenFeature, ProviderEvents, InMemoryProvider } from '../../src';
```

**Comment by toddbaert** [comment]
So the `evaluation.feature` gherkin file is currently being supplied from the `test-harness` module, but again, in our efforts to remove all flagd-related dependencies from the SDKs, this gherkin is now available in the [spec](https://github.com/open-feature/spec/blob/main/specification/appendix-b-gherkin-suites.md) (with a few minor changes).

I've made a [new issue](https://github.com/open-feature/js-sdk/issues/743) to replace the `test-harness` submodule with a new spec submodule that will supply this gherkin.

**Review by toddbaert — APPROVED** [review]
Approved see comment [here](https://github.com/open-feature/js-sdk/pull/740#issuecomment-1879030210) about submodule.

@beeme1mr feel free to merge.

**Review by lukas-reining — APPROVED** [review]

**Pull request merged**